src/jobs.rs

```diff
@@ -0,0 +1,51 @@
+//! SCHEDULED JOBS
+//!
+//! The metadata is a serde_json::Value
+//! Please refer to https://docs.rs/serde_json/latest/serde_json/value/fn.from_value.html
+//! on how to interpret it as an instance of type T, implementing Serialize/Deserialize.
+//!
+//! The schedule is a cron::Schedule
+//! Please refer to https://docs.rs/cron/latest/cron/struct.Schedule.html for further info
+//!
+//! For example, if we want to sends a Zulip message every Friday at 11:30am ET into #t-release
+//! with a @T-release meeting! content, we should create some JobSchedule like:
+//!
+//!    #[derive(Serialize, Deserialize)]
+//!    struct ZulipMetadata {
+//!      pub message: String
+//!    }
+//!
+//!    let metadata = serde_json::value::to_value(ZulipMetadata {
+//!      message: "@T-release meeting!".to_string()
+//!    }).unwrap();
+//!
+//!    let schedule = Schedule::from_str("0 30 11 * * FRI *").unwrap();
+//!    
+//!    let new_job = JobSchedule {
+//!      name: "send_zulip_message".to_owned(),
+//!      schedule: schedule,
+//!      metadata: metadata
+//!    }
+//!
+//! and include it in the below vector in jobs():
+//!
+//!   jobs.push(new_job);
+//!
+//! ... fianlly, add the corresponding "send_zulip_message" handler in src/handlers/jobs.rs
+
+use crate::db::jobs::JobSchedule;
+
+// How often new cron-based jobs will be placed in the queue.
+// This is the minimum period *between* a single cron task's executions.
+pub const JOB_SCHEDULING_CADENCE_IN_SECS: u64 = 1800;
+
+// How often the database is inspected for jobs which need to execute.
+// This is the granularity at which events will occur.
+pub const JOB_PROCESSING_CADENCE_IN_SECS: u64 = 60;
```

Comment:
what happens if a job (for any reason) end up taking longer than `JOB_PROCESSING_CADENCE_IN_SECS`?

Reply:
Then that could cause other scheduled jobs (if any for the next interval) to be executed later than expected. 
We did worry about this earlier and thought about processing each job asynchronously,  but got a comment from @Mark-Simulacrum (somewhere above) saying that:
> This is fine. We shouldn't really have long-running tasks, so awaiting inline seems fine to me.

perhaps we could run `db::schedule_jobs` async (from a tokio task) inside the loop to not potentially block/delay the next interval?

Reply:
Yeah, I wouldn't worry about this. In an ideal world we'd be emitting a metric here for each task's duration and alarming on it if it hits >5 seconds, say, but we're not really set up for that today so for now we'll just leave this as-is.

I wouldn't expect us to have more than one task in any given job execution point essentially ever, so this is pretty unlikely to become a problem. And most tasks will be "post a comment", which takes ~milliseconds.

Reply:
ok thanks for rebulding the discussion for me! I had lost it in the web interface of github :)